contracts/tips/src/msg.rs

```diff
@@ -0,0 +1,266 @@
+use crate::error::ContractError;
+use crate::state::StateServiceFee;
+use cosmwasm_std::{Addr, Coin, Decimal, Uint64};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+/// Fees required to execute [`ExecuteMsg::SendTip`].
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ServiceFee {
+    /// Represents a fixed fee that the sender needs to pay in order to send the tip.
+    Fixed {
+        /// Coins that the sender needs to pay.
+        amount: Vec<Coin>,
+    },
+    /// Represents a percentage that the sender needs to pay in order to send the tip.
+    Percentage {
+        /// Percentage value.
+        value: Decimal,
+    },
+}
+
+impl ServiceFee {
+    pub fn validate(&self) -> Result<(), ContractError> {
+        match self {
+            ServiceFee::Fixed { amount } => {
+                if amount.is_empty() {
+                    return Err(ContractError::EmptyFixedFee {});
+                }
+
+                let zero_coin = amount.iter().find(|coin| coin.amount.is_zero());
+                if let Some(coin) = zero_coin {
+                    return Err(ContractError::ZeroFeeCoin {
+                        denom: coin.denom.to_owned(),
+                    });
+                }
+            }
+            ServiceFee::Percentage { value } => {
+                let one_hundred = Decimal::from_atomics(100u32, 0).unwrap();
+                if value.ge(&one_hundred) || value.is_zero() {
+                    return Err(ContractError::InvalidPercentageFee {});
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl From<StateServiceFee> for ServiceFee {
+    fn from(state_service_fees: StateServiceFee) -> Self {
+        match state_service_fees {
+            StateServiceFee::Fixed { amount } => ServiceFee::Fixed { amount },
+            StateServiceFee::Percentage { value } => ServiceFee::Percentage { value },
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct InstantiateMsg {
+    /// Address of who will have the right to manage the contract.
+    pub admin: String,
+    /// Application which is deploying the contract.
+    pub subspace_id: Uint64,
+    /// Fee that the users need to pay to use the contract.
+    /// If `None` no fees will be collected from the tipped amount.
+    pub service_fee: Option<ServiceFee>,
+    /// The number of records saved of a user tips history.
+    pub saved_tips_threshold: u32,
```

Comment:
I would name it `saved_tips_size`. Threshold to me is something that the lowest condition to save.